File: packages/julia/src/Seguranca/Ataques/AtaqueCoberturaDependenciaIV.jl
"""
Complemento do AtaqueCoberturaDependencia (que varia a CHAVE): aqui varia
o IV. Toda posição do IV precisa influenciar toda posição de saída; uma
posição "morta" do IV reduziria a entropia efetiva que o IV injeta.
"""
struct AtaqueCoberturaDependenciaIV <: Ataque
    tamanho_bloco::Int
    perturbacoes_por_posicao::Int
end

AtaqueCoberturaDependenciaIV(;
    tamanho_bloco::Integer = 32,
    perturbacoes_por_posicao::Integer = 5,
) = AtaqueCoberturaDependenciaIV(Int(tamanho_bloco), Int(perturbacoes_por_posicao))

nome(::AtaqueCoberturaDependenciaIV) = "Cobertura de dependência do IV (entrada x saída)"

function executar(a::AtaqueCoberturaDependenciaIV, alvo::AlvoCriptografico)
    chave = chave_de_teste(alvo)
    tam_iv = tamanho_iv(alvo)
    iv_base = zeros(UInt8, tam_iv)

    ks_base = gerar_keystream_bruto(alvo, chave, iv_base, "enc", a.tamanho_bloco)

    pares_independentes = String[]

    for pos_iv in 0:(tam_iv-1)
        afetou = falses(a.tamanho_bloco)

        for _ in 1:(a.perturbacoes_por_posicao)
            iv = copy(iv_base)
            iv[pos_iv+1] = UInt8(random_int(1, 255))
            ks = gerar_keystream_bruto(alvo, chave, iv, "enc", a.tamanho_bloco)

            for pos_saida in 0:(a.tamanho_bloco-1)
                if ks[pos_saida+1] != ks_base[pos_saida+1]
                    afetou[pos_saida+1] = true
                end
            end
        end

        for pos_saida in 0:(a.tamanho_bloco-1)
            if !afetou[pos_saida+1]
                push!(pares_independentes, "iv[$pos_iv] -> saida[$pos_saida]")
            end
        end
    end

    if !isempty(pares_independentes)
        return ResultadoAtaque(
            nome(a),
            true,
            "alta",
            "$(length(pares_independentes)) par(es) sem dependência detectável em $(a.perturbacoes_por_posicao) tentativas cada",
            Dict{String,Any}(
                "pares" => pares_independentes[1:min(20, length(pares_independentes))],
            ),
        )
    end

    return ResultadoAtaque(
        nome(a),
        false,
        "info",
        "Todas as $tam_iv posições do IV influenciam todas as $(a.tamanho_bloco) posições de saída",
    )
end

File: packages/rust/src/seguranca/ataques/ataque_cobertura_dependencia_iv.rs
//! Complemento do AtaqueCoberturaDependencia (que varia a CHAVE): aqui varia
//! o IV. Toda posição do IV precisa influenciar toda posição de saída; uma
//! posição "morta" do IV reduziria a entropia efetiva que o IV injeta.

use crate::seguranca::alvo_criptografico::AlvoCriptografico;
use crate::seguranca::ataque::{Ataque, ErroAtaque};
use crate::seguranca::resultado_ataque::{ResultadoAtaque, Severidade};
use crate::seguranca::util::random_int;

pub struct AtaqueCoberturaDependenciaIV {
    pub tamanho_bloco: usize,
    pub perturbacoes_por_posicao: usize,
}

impl AtaqueCoberturaDependenciaIV {
    pub fn new() -> Self {
        AtaqueCoberturaDependenciaIV {
            tamanho_bloco: 32,
            perturbacoes_por_posicao: 5,
        }
    }
}

impl Default for AtaqueCoberturaDependenciaIV {
    fn default() -> Self {
        Self::new()
    }
}

impl Ataque for AtaqueCoberturaDependenciaIV {
    fn nome(&self) -> String {
        "Cobertura de dependência do IV (entrada x saída)".to_string()
    }

    fn executar(&self, alvo: &dyn AlvoCriptografico) -> Result<ResultadoAtaque, ErroAtaque> {
        let chave = alvo.chave_de_teste();
        let tamanho_iv = alvo.tamanho_iv();
        let iv_base = vec![0u8; tamanho_iv];

        let ks_base =
            alvo.gerar_keystream_bruto(chave.as_bytes(), &iv_base, "enc", self.tamanho_bloco);

        let mut pares_independentes: Vec<String> = Vec::new();

        for pos_iv in 0..tamanho_iv {
            let mut afetou = vec![false; self.tamanho_bloco];

            for _ in 0..self.perturbacoes_por_posicao {
                let mut iv = iv_base.clone();
                iv[pos_iv] = random_int(1, 255) as u8;
                let ks = alvo.gerar_keystream_bruto(chave.as_bytes(), &iv, "enc", self.tamanho_bloco);

                for pos_saida in 0..self.tamanho_bloco {
                    if ks[pos_saida] != ks_base[pos_saida] {
                        afetou[pos_saida] = true;
                    }
                }
            }

            for (pos_saida, &ok) in afetou.iter().enumerate() {
                if !ok {
                    pares_independentes
                        .push(format!("iv[{}] -> saida[{}]", pos_iv, pos_saida));
                }
            }
        }

        if !pares_independentes.is_empty() {
            return Ok(ResultadoAtaque::new(
                &self.nome(),
                true,
                Severidade::Alta,
                &format!(
                    "{} par(es) sem dependência detectável em {} tentativas cada",
                    pares_independentes.len(),
                    self.perturbacoes_por_posicao
                ),
            ));
        }

        Ok(ResultadoAtaque::new(
            &self.nome(),
            false,
            Severidade::Info,
            &format!(
                "Todas as {} posições do IV influenciam todas as {} posições de saída",
                tamanho_iv, self.tamanho_bloco
            ),
        ))
    }
}
